addons/mail_client_extension/controllers/mail_client_extension.py

```diff
@@ -14,32 +15,159 @@
+
+        normalized_email = partner.email_normalized
+        if not normalized_email:
+            response = {'error': 'contact has no email'}
+            return response
+
+        company_domain = normalized_email.split('@')[1]
+
+        company, enrichment_info = self._create_company_from_iap(company_domain)
+
+        response['enrichment_info'] = enrichment_info
+        response['company'] = self._prepare_company_values(company)
+        if company:
+            partner.write({'parent_id': company})
+
+        return response
+
+    @http.route('/mail_client_extension/partner/get', type="json", auth="outlook", cors="*")
+    def res_partner_get(self, email=None, name=None, partner_id=None, **kwargs):
+        """
+        returns a partner given it's id or an email and a name.
+        In case the partner does not exist, we return partner having an id -1, we also look if an existing company
+        matching the contact exists in the database, if none is found a new company is enriched and created automatically
+        """
+        response = {}
+
+        if not (partner_id or (name and email)):
+            return {'error': _('You need to specify at least the partner_id or the name and the email')}
+
+        if partner_id:
+            partner = request.env['res.partner'].browse(partner_id)
+            return self._prepare_contact_values(partner)
+
+        normalized_email = tools.email_normalize(email)
+        if not normalized_email:
+            response['error'] = 'Bad email.'
+            return response
+
+        # Search for the partner based on the email.
+        # If multiple are found, take the first one.
+        partner = request.env['res.partner'].search(['|', ('email', 'in', [normalized_email, email]),
+                                                     ('email_normalized', '=', normalized_email)], limit=1)
+
+        response = self._prepare_contact_values(partner)
+
+        # if no partner is found in the database, we should also return an empty one having id = -1, otherwise older versions of
+        # plugin won't work
+        if not response['partner']:
+            sender_domain = normalized_email.split('@')[1]
+            response['partner'] = {
+                'id': -1,
+                'email': email,
+                'name': name,
+                'enrichment_info': None,
+            }
+            company = self._find_existing_company(normalized_email)
+            if not company:  # create and enrich company
+                company, enrichment_info = self._create_company_from_iap(sender_domain)
+                response['partner']['enrichment_info'] = enrichment_info
+            response['partner']['company'] = self._prepare_company_values(company)
+
+        return response
+
+    @http.route('/mail_client_extension/partners/search', type="json", auth="outlook", cors="*")
+    def res_partners_search(self, query, limit=30, **kwargs):
```

Comment:
query generally indicates an sql_query in odoo... I would use search_term ou partner_name maybe ?

Reply:
Changed to "search_term" as "partner_name" can be misleading since we are also performing searches on emails and refs.

@std-odoo, this would also require some light changes in the gmail plugin :)